source/common/common/mutex_tracer.cc

```diff
@@ -0,0 +1,39 @@
+#include "common/common/mutex_tracer.h"
+
+#include <iostream>
+#include <memory>
+
+#include "common/common/assert.h"
+
+#include "absl/synchronization/mutex.h"
+
+namespace Envoy {
+
+MutexTracer* MutexTracer::singleton_ = nullptr;
+
+MutexTracer* MutexTracer::getOrCreateTracer() {
+  if (singleton_ == nullptr) {
+    singleton_ = new MutexTracer;
+    absl::RegisterMutexTracer(&Envoy::MutexTracer::contentionHook);
+  }
+  return singleton_;
+}
+
+void MutexTracer::contentionHook(const char* msg, const void* obj, int64_t wait_cycles) {
```

Comment:
Yeah, the `obj` arg is an "opaque handle to the contended mutex" per the [docs](https://github.com/abseil/abseil-cpp/blob/cc8dcd307b76a575d2e3e0958a4fe4c7193c2f68/absl/synchronization/mutex.h#L944-L954). It would be very neat to track not just contention but keep a (histogram?) list of which mutexes where for debugging.